File: python/src/caspian/core/overlap.py
"""Overlap FSM — pure state transitions, no asyncio.Queue, no I/O.

The overlap policy is named on the Rule. The state machine computes transitions.
The runner (memory / process / hosted) executes the actual queuing.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from caspian.core.types import OverlapPolicy


class SlotStatus(StrEnum):
    IDLE = "idle"
    BUSY = "busy"


class OverlapState(BaseModel):
    """Per-key overlap state. Managed by the runner, transitions computed by core."""

    model_config = ConfigDict(frozen=True, extra="forbid")
    status: SlotStatus = SlotStatus.IDLE
    queued: int = 0
    skipped_count: int = 0


class OverlapDecision(StrEnum):
    EXECUTE = "execute"
    ENQUEUE = "enqueue"
    DROP = "drop"


class OverlapResult(BaseModel):
    """Result of an overlap transition."""

    model_config = ConfigDict(frozen=True, extra="forbid")
    decision: OverlapDecision
    new_state: OverlapState
    skipped_count: int = 0


def overlap_transition(
    state: OverlapState,
    policy: OverlapPolicy,
    bound: int,
) -> OverlapResult:
    """Pure transition: given current state + policy, decide what to do with a new event.

    Returns the decision (execute/enqueue/drop) and the new state.
    """
    match policy:
        case OverlapPolicy.PARALLEL:
            return OverlapResult(
                decision=OverlapDecision.EXECUTE,
                new_state=state,
            )

        case OverlapPolicy.DROP:
            if state.status == SlotStatus.BUSY:
                return OverlapResult(
                    decision=OverlapDecision.DROP,
                    new_state=state,
                )
            return OverlapResult(
                decision=OverlapDecision.EXECUTE,
                new_state=OverlapState(status=SlotStatus.BUSY),
            )

        case OverlapPolicy.QUEUE:
            if state.status == SlotStatus.IDLE:
                return OverlapResult(
                    decision=OverlapDecision.EXECUTE,
                    new_state=OverlapState(status=SlotStatus.BUSY, queued=0),
                )
            if state.queued >= bound:
                return OverlapResult(
                    decision=OverlapDecision.DROP,
                    new_state=state,
                )
            return OverlapResult(
                decision=OverlapDecision.ENQUEUE,
                new_state=OverlapState(
                    status=SlotStatus.BUSY,
                    queued=state.queued + 1,
                    skipped_count=state.skipped_count + 1,
                ),
            )

        case OverlapPolicy.DEBOUNCE:
            if state.status == SlotStatus.IDLE:
                return OverlapResult(
                    decision=OverlapDecision.EXECUTE,
                    new_state=OverlapState(status=SlotStatus.BUSY),
                )
            return OverlapResult(
                decision=OverlapDecision.ENQUEUE,
                new_state=OverlapState(
                    status=SlotStatus.BUSY,
                    queued=1,
                    skipped_count=state.skipped_count + 1,
                ),
            )

        case _:
            return OverlapResult(
                decision=OverlapDecision.EXECUTE,
                new_state=state,
            )


def drain_transition(state: OverlapState, policy: OverlapPolicy) -> OverlapResult:
    """Called when a handler completes. Decides whether to run the next queued event."""
    if state.queued > 0 and policy in (OverlapPolicy.QUEUE, OverlapPolicy.DEBOUNCE):
        return OverlapResult(
            decision=OverlapDecision.EXECUTE,
            new_state=OverlapState(
                status=SlotStatus.BUSY,
                queued=0,
                skipped_count=0,
            ),
            skipped_count=state.skipped_count,
        )
    return OverlapResult(
        decision=OverlapDecision.DROP,
        new_state=OverlapState(status=SlotStatus.IDLE),
    )

File: packages/typescript/src/core/overlap.ts
/**
 * Overlap FSM — pure transitions, no queues, no I/O.
 *
 * The rule names the policy. This module only computes the next state.
 * The runner owns the actual wait/buffer.
 */
import * as Schema from "effect/Schema"
import type { OverlapPolicy } from "./app.ts"

export const SlotStatus = Schema.Literal("idle", "busy")
export type SlotStatus = typeof SlotStatus.Type

export const OverlapState = Schema.Struct({
  status: SlotStatus,
  queued: Schema.Number.pipe(Schema.int(), Schema.greaterThanOrEqualTo(0)),
  skipped_count: Schema.Number.pipe(Schema.int(), Schema.greaterThanOrEqualTo(0)),
})
export type OverlapState = typeof OverlapState.Type

export const OverlapDecision = Schema.Literal("execute", "enqueue", "drop")
export type OverlapDecision = typeof OverlapDecision.Type

export const OverlapResult = Schema.Struct({
  decision: OverlapDecision,
  new_state: OverlapState,
  skipped_count: Schema.Number.pipe(Schema.int(), Schema.greaterThanOrEqualTo(0)),
})
export type OverlapResult = typeof OverlapResult.Type

export const idleOverlapState: OverlapState = {
  status: "idle",
  queued: 0,
  skipped_count: 0,
}

const busy = (queued: number, skipped_count: number): OverlapState => ({
  status: "busy",
  queued,
  skipped_count,
})

const result = (
  decision: OverlapDecision,
  new_state: OverlapState,
  skipped_count = 0,
): OverlapResult => ({ decision, new_state, skipped_count })

export const overlapTransition = (
  state: OverlapState,
  policy: OverlapPolicy,
  bound: number,
): OverlapResult => {
  switch (policy) {
    case "parallel":
      return result("execute", state)
    case "drop":
      if (state.status === "busy") {
        return result("drop", state)
      }
      return result("execute", busy(0, 0))
    case "queue":
      if (state.status === "idle") {
        return result("execute", busy(0, 0))
      }
      if (state.queued >= bound) {
        return result("drop", state)
      }
      return result(
        "enqueue",
        busy(state.queued + 1, state.skipped_count + 1),
      )
    case "debounce":
      if (state.status === "idle") {
        return result("execute", busy(0, 0))
      }
      return result("enqueue", busy(1, state.skipped_count + 1))
  }
}

export const drainTransition = (
  state: OverlapState,
  policy: OverlapPolicy,
): OverlapResult => {
  if (state.queued > 0 && (policy === "queue" || policy === "debounce")) {
    return result("execute", busy(0, 0), state.skipped_count)
  }
  return result("drop", idleOverlapState)
}
